Neos.ContentRepositoryRegistry/Classes/Service/CatchUpDeduplicationQueue.php

```diff
@@ -0,0 +1,171 @@
+<?php
+namespace Neos\ContentRepositoryRegistry\Service;
+
+use Neos\Cache\Frontend\FrontendInterface;
+use Neos\ContentRepository\Core\Factory\ContentRepositoryId;
+use Neos\ContentRepository\Core\Projection\ProjectionCatchUpTriggerInterface;
+use Neos\ContentRepository\Core\Projection\Projections;
+
+/**
+ *
+ */
+final readonly class CatchUpDeduplicationQueue
+{
+    public function __construct(
+        private ContentRepositoryId $contentRepositoryId,
+        private FrontendInterface $catchUpLock,
+        private ProjectionCatchUpTriggerInterface $projectionCatchUpTrigger
+    ) {}
+
+    public function requestCatchUp(Projections $projections): void
+    {
+        $queuedProjections = $this->triggerCatchUpAndReturnQueued($projections);
+        $attempts = 0;
+        /** @phpstan-ignore-next-line */
+        while ($queuedProjections->isEmpty() === false) {
+            usleep(random_int(100, 100 + $attempts) + ($attempts * $attempts * 10));
+            $queuedProjections = $this->retryQueued($queuedProjections);
+            $attempts++;
```

Comment:
There is no timeout / max attempts.. that makes me suspicious..

Reply:
I had that before but reasoning with myself says there shouldn't be any, we can't know how long the catchup takes, but we do know that we wanted another one, so we just got to wait...

Reply:
I did a quick test using the amp package to make this wait async, that works well, but I would do that in a follow up.

Reply:
I used [react/promise](https://reactphp.org/promise/) for that in my custom implementation and I really like the simple way to use it:

```php
$this->catchUpTrigger->start()
    ->then(
        fn (SequenceNumber $sequenceNumber) => $this->foo($sequenceNumber),
        fn (Throwable $exception) => $this->bar($exception)
    );
```

(but AMP is probably just the same)

Still, I think it would be dangerous to kill the server if some projection hangs and would rather stop with a meaningful exception after a reasonable amount of time

Reply:
The question is, what is a reasonable amount of time, if a long catchup is running (eg after a big publish) it can take quite a while to get your queued catchup in. 

ASIDE that makes the bottom maybe useless, just came to my mind... What about this:
Now that we have this Queue also in the EventPersister to unlock after, we might as well defer the retriggering to after a catchup happened?! That way we could totally skip the loop here and just do something like

```
        $catchUp->run($eventStream);
        $catchUpHook?->onAfterCatchUp();
        $this->catchUpDeduplicationQueue->releaseCatchUpLock($projectionClassName);
        $this->catchUpDeduplicationQueue->startIfQueued($projectionClassName);
```


Note that you have two ways to get out of the loop, not only by starting another catchup there but also if it was dequeued due to another process starting a catchup, so IMHO it should be fine to risk, but we can obivously add a generous end condition, any suggestions? 1 minute ? 5?

Reply:
Means we only really block if we are forced by the `block()`  thing.

Reply:
> we might as well defer the retriggering to after a catchup happened?!

We thought about this before, but this could mean that one process has to wait endlessly (not very likely, but in a scenario where many other systems interact with the CR).
That's why we decided to put the "burden" on to the invoking side of the new catch up (that now has to wait for the first catch up to finish)

Reply:
I agree, sth like 20 seconds I'd say.
Remember: This won't intercept the catchup or start a new one if the time exeeded. It just avoids that a process waits for other catchups endlessly

Reply:
Here we go, with a bit of explainer added as well.

Reply:
I realized, that we can't use the PSR cache interfaces for this because they don't allow for exclusive cache entries. Instead we'll need some lock (e.g. [symfony/lock](https://symfony.com/doc/current/components/lock.html)) to prevent race conditions.
Thanks to the nice abstraction, it should be relatively easy to replace the cache.

I'll experiment a bit with the lock and ways to test this for reliability & concurrency and post an update here asap

Reply:
Yeah, this is not atomic, so it's not a 100% failsafe and for sure th more pressure the more duplicate runs probably can happen... I am not sure if it still isn't better than what we have now and we might merge this and refactor to lock or similar afterwards, but I am fine either way.

Reply:
Note: I was on my smartphone and actually meant to comment on the PR itself, not on this particular code

> more pressure the more duplicate runs probably can happen

The problem is that the same event could be consumed multiple times by a single listener – it is currently prevented by our `CheckpointStorage` implementation, but I would really like to simplify that one.
I wrote a simplified example with a test using [paratest](https://github.com/paratestphp/paratest) and it failed every time – using the lock seemed faster and the race condition was gone.

There are also two more details that popped up.. Are you up to discuss those at some point?